File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/serialization/MapProvider.java
package io.github.romangolovanov.apps.ametro.model.serialization;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;

import io.github.romangolovanov.apps.ametro.model.entities.MapLocale;
import io.github.romangolovanov.apps.ametro.model.entities.MapMetadata;
import io.github.romangolovanov.apps.ametro.model.entities.MapScheme;
import io.github.romangolovanov.apps.ametro.model.entities.MapStationInformation;
import io.github.romangolovanov.apps.ametro.model.entities.MapTransportScheme; /**
 * Abstraction for different sources of map content (zip archive, assets folder, etc.)
 */
public interface MapProvider extends AutoCloseable {

    String[] getSupportedLocales() throws IOException;

    HashMap<Integer, String> getTextsMap(String languageCode) throws IOException;

    HashMap<Integer, List<String>> getAllTextsMap() throws IOException;

    MapMetadata getMetadata(MapLocale locale) throws IOException;

    MapTransportScheme getTransportScheme(String name) throws IOException;

    MapScheme getScheme(String name, MapLocale locale) throws IOException;

    Object getBackgroundObject(String name) throws IOException;

    MapStationInformation[] getStationInformation() throws IOException;

    String getFileContent(String name) throws IOException;

    @Override
    void close() throws IOException;
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/serialization/MapProvider.kt
package io.github.romangolovanov.apps.ametro.model.serialization

import java.io.IOException

import io.github.romangolovanov.apps.ametro.model.entities.MapLocale
import io.github.romangolovanov.apps.ametro.model.entities.MapMetadata
import io.github.romangolovanov.apps.ametro.model.entities.MapScheme
import io.github.romangolovanov.apps.ametro.model.entities.MapStationInformation
import io.github.romangolovanov.apps.ametro.model.entities.MapTransportScheme

/**
 * Abstraction for different sources of map content (zip archive, assets folder, etc.)
 */
interface MapProvider : AutoCloseable {
    @Throws(IOException::class)
    fun getSupportedLocales(): Array<String>

    @Throws(IOException::class)
    fun getTextsMap(languageCode: String): Map<Int, String>

    @Throws(IOException::class)
    fun getAllTextsMap(): Map<Int, MutableList<String>>

    @Throws(IOException::class)
    fun getMetadata(locale: MapLocale?): MapMetadata

    @Throws(IOException::class)
    fun getTransportScheme(name: String): MapTransportScheme

    @Throws(IOException::class)
    fun getScheme(name: String, locale: MapLocale): MapScheme

    @Throws(IOException::class)
    fun getBackgroundObject(name: String): Any

    @Throws(IOException::class)
    fun getStationInformation(): Array<MapStationInformation>

    @Throws(IOException::class)
    fun getFileContent(name: String): String

    @Throws(IOException::class)
    override fun close()
}
